File: bms/go/geo/geo_s1_Interval_ApproxEqual/geo_s1_Interval_ApproxEqual.go
package test

import (
	"math"
)

var (
	geo_s1_epsilon = 1e-15
)

func (i geo_s1_Interval) ApproxEqual(other geo_s1_Interval) bool {
	// Full and empty intervals require special cases because the endpoints
	// are considered to be positioned arbitrarily.
	if i.IsEmpty() {
		return other.Length() <= 2*geo_s1_epsilon
	}
	if other.IsEmpty() {
		return i.Length() <= 2*geo_s1_epsilon
	}
	if i.IsFull() {
		return other.Length() >= 2*(math.Pi-geo_s1_epsilon)
	}
	if other.IsFull() {
		return i.Length() >= 2*(math.Pi-geo_s1_epsilon)
	}

	// The purpose of the last test below is to verify that moving the endpoints
	// does not invert the interval, e.g. [-1e20, 1e20] vs. [1e20, -1e20].
	return (math.Abs(math.Remainder(other.Lo-i.Lo, 2*math.Pi)) <= geo_s1_epsilon &&
		math.Abs(math.Remainder(other.Hi-i.Hi, 2*math.Pi)) <= geo_s1_epsilon &&
		math.Abs(i.Length()-other.Length()) <= 2*geo_s1_epsilon)

}

func (i geo_s1_Interval) IsEmpty() bool { return i.Lo == math.Pi && i.Hi == -math.Pi }

func (i geo_s1_Interval) Length() float64 {
	l := i.Hi - i.Lo
	if l >= 0 {
		return l
	}
	l += 2 * math.Pi
	if l > 0 {
		return l
	}
	return -1
}

func (i geo_s1_Interval) IsFull() bool { return i.Lo == -math.Pi && i.Hi == math.Pi }

type geo_s1_Interval struct {
	Lo, Hi float64
}

File: translations_only/fail/go/geo/geo_s1_Interval_ApproxEqual/claude2/Hinted/geo_s1_Interval_ApproxEqual.rs

use std::f64::consts::PI;

const GEO_S1_EPSILON: f64 = 1e-15;

fn approx_equal(i: &Interval, other: &Interval) -> bool {
    // Full and empty intervals require special cases because the endpoints
    // are considered to be positioned arbitrarily.
    if is_empty(i) {
        return length(other) <= 2.0 * GEO_S1_EPSILON;
    }
    if is_empty(other) {
        return length(i) <= 2.0 * GEO_S1_EPSILON;
    }
    if is_full(i) {
        return length(other) >= 2.0 * (PI - GEO_S1_EPSILON);
    }
    if is_full(other) {
        return length(i) >= 2.0 * (PI - GEO_S1_EPSILON);
    }

    // The purpose of the last test below is to verify that moving the endpoints
    // does not invert the interval, e.g. [-1e20, 1e20] vs. [1e20, -1e20].
    (f64::abs(f64::rem_euclid(other.lo - i.lo, 2.0 * PI)) <= GEO_S1_EPSILON)
        && (f64::abs(f64::rem_euclid(other.hi - i.hi, 2.0 * PI)) <= GEO_S1_EPSILON)
        && (f64::abs(length(i) - length(other)) <= 2.0 * GEO_S1_EPSILON)
}

fn is_empty(i: &Interval) -> bool {
    i.lo == PI && i.hi == -PI 
}

fn length(i: &Interval) -> f64 {
    let l = i.hi - i.lo;
    if l >= 0.0 {
        l
    } else {
        l + 2.0 * PI
    }.max(0.0)
}

fn is_full(i: &Interval) -> bool {
    i.lo == -PI && i.hi == PI
}

struct Interval {
    lo: f64,
    hi: f64
}

